### PR TITLE
[UAR-137] WIP: check your answers lauches createPayment()

### PR DESCRIPTION
**Pull request by corybudd-ch**
### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/update/check.your.answers.controller.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+export const post = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    logger.debugRequest(req, `POST ${UPDATE_CHECK_YOUR_ANSWERS_PAGE}`);
+
+    const session = req.session as Session;
+    const appData: ApplicationData = getApplicationData(session);
+
+    let transactionID, overseasEntityID;
+    if (isActiveFeature(FEATURE_FLAG_ENABLE_SAVE_AND_RESUME_17102022)) {
```
Comment (AndreKramer7654) [inline src/controllers/update/check.your.answers.controller.ts:47]: We should add 
 if (!isActiveFeature(config.FEATURE_FLAG_ENABLE_ROE_UPDATE) && isActiveFeature(config.FEATURE_FLAG_ENABLE_SAVE_AND_RESUME_17102022)) {

Comment (corybudd-ch) [reply src/controllers/update/check.your.answers.controller.ts:47]: I'm just thinking should we even have this `if` block at all? Just keep the `else` part?
 Since there is the existing controller for the register journey, and this new controller will only be used for update journey this code behind `!isActiveFeature(config.FEATURE_FLAG_ENABLE_ROE_UPDATE) && isActiveFeature(config.FEATURE_FLAG_ENABLE_SAVE_AND_RESUME_17102022))` wouldn't get called

File: src/service/payment.service.ts [hunk 2]
```diff
@@ -72,11 +72,15 @@ export const startPaymentsSession = async (
...
-const setPaymentRequest = (transactionId: string, overseasEntityId: string): CreatePaymentRequest => {
+const setPaymentRequest = (transactionId: string, overseasEntityId: string, baseURL?: string): CreatePaymentRequest => {
 
   const paymentResourceUri = `${API_URL}/transactions/${transactionId}/${PAYMENT}`;
 
-  const baseURL = `${CHS_URL}${REGISTER_AN_OVERSEAS_ENTITY_URL}`;
+  if (!baseURL) {
+    baseURL = `${CHS_URL}${REGISTER_AN_OVERSEAS_ENTITY_URL}`;
+  }
+  logger.debug("AKDEBUG payment return baseURL " + baseURL);
```
Comment (stephanie-ch) [inline src/service/payment.service.ts:82]: Is this debug line needed anymore?

File: src/controllers/update/check.your.answers.controller.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
+import { NextFunction, Request, Response } from "express";
+
+import { logger } from "../../utils/logger";
+import { getApplicationData } from "../../utils/application.data";
```
Comment (stephanie-ch) [inline src/controllers/update/check.your.answers.controller.ts:4]: I checked with @corybudd-ch and the project uses these relative paths elsewhere but it's fairly bad practice. It might be worth seeing about adding a tech debt ticket to sort the project out with absolute paths?

File: src/controllers/update/check.your.answers.controller.ts (hunk 1 above)
Comment (corybudd-ch) [reply src/controllers/update/check.your.answers.controller.ts]: I've simplified the code block for now if there is upcoming save & resume epic.

File: test/controllers/update/check.your.answers.controller.spec.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  });
+});
+
+describe("POST tests", () => {
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  test(`redirect to ${PAYMENT_LINK_JOURNEY}, the first Payment web journey page, after fetching transaction and OE id from appData`, async () => {
```
Comment (stephanie-ch) [inline test/controllers/update/check.your.answers.controller.spec.ts:82]: Minor - The test name might be a bit wordy. You can say the same with "redirect to initial page of payment journey with transaction and OE ID"

File: test/controllers/update/check.your.answers.controller.spec.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    mockGetApplicationData.mockReturnValueOnce(APPLICATION_DATA_MOCK);
+    mockPaymentsSession.mockReturnValueOnce(PAYMENT_LINK_JOURNEY);
+    const resp = await request(app).post(UPDATE_CHECK_YOUR_ANSWERS_URL);
+
+    expect(resp.status).toEqual(302);
+    expect(resp.text).toEqual(`${FOUND_REDIRECT_TO} ${PAYMENT_LINK_JOURNEY}`);
+  });
+
+  test(`catch error when post data from ${UPDATE_CHECK_YOUR_ANSWERS_PAGE} page`, async () => {
```
Comment (stephanie-ch) [inline test/controllers/update/check.your.answers.controller.spec.ts:91]: Catch error when post data what from page? Perhaps this name needs some clarification - It made my brain hurt :)

File: views/update/update-check-your-answers.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
+{% extends "layout.html" %}
+
+{% block pageTitle %}
+  Check your answers before completing your update - {{ SERVICE_NAME }} - GOV.UK
```
Comment (stephanie-ch) [inline views/update/update-check-your-answers.html:4]: Is all text confirmed by a content designer?

File: views/update/update-check-your-answers.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+{% block backLink %}
+  {% include "includes/back-link.html" %}
+{% endblock %}
+
+{% block content %}
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds-from-desktop">
+
+      <h1 class="govuk-heading-xl">Check your answers before sending your application (NOT LIVE)</h1>
```
Comment (stephanie-ch) [inline views/update/update-check-your-answers.html:15]: Do we want to see "NOT LIVE" when the update flag is enabled?

**Review by stephanie-ch — CHANGES_REQUESTED** [review]

File: views/update/update-check-your-answers.html (hunk 6 above)
Comment (corybudd-ch) [reply views/update/update-check-your-answers.html:4]: We have no content for any of our pages from designers yet :( so all our tickets are just implementing basic pages and the functionality with the idea, I believe, that the content etc will all be finalised before it goes live.

File: views/update/update-check-your-answers.html (hunk 7 above)
Comment (corybudd-ch) [reply views/update/update-check-your-answers.html:15]: Same as the other comment, I vaguely remember either a content designer or someone from product saying if we can mark all our pages as NOT LIVE in some way so this is what we've been doing with all of them so far. Then we'll update before all this goes live.

File: src/service/payment.service.ts (hunk 2 above)
Comment (corybudd-ch) [reply src/service/payment.service.ts]: removed

File: test/controllers/update/check.your.answers.controller.spec.ts (hunk 4 above)
Comment (corybudd-ch) [reply test/controllers/update/check.your.answers.controller.spec.ts]: updated

File: test/controllers/update/check.your.answers.controller.spec.ts (hunk 5 above)
Comment (corybudd-ch) [reply test/controllers/update/check.your.answers.controller.spec.ts]: updated

**Review by AndreKramer7654 — APPROVED** [review]

**Review by cdrainCH — APPROVED** [review]

**Review by AndreKramer7654 — APPROVED** [review]

**Review by stephanie-ch — APPROVED** [review]

File: src/controllers/update/check.your.answers.controller.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import {
+  CHS_URL,
+  OVERSEAS_ENTITY_QUERY_URL,
+  UPDATE_AN_OVERSEAS_ENTITY_URL,
+  UPDATE_CHECK_YOUR_ANSWERS_PAGE
+} from "../../config";
+
+export const get = (req: Request, res: Response, next: NextFunction) => {
+  try {
+    logger.debugRequest(req, `GET ${UPDATE_CHECK_YOUR_ANSWERS_PAGE}`);
```
Comment (dgolubevs) [inline src/controllers/update/check.your.answers.controller.ts:21]: better (same for POST)
```
 logger.debugRequest(req, `${req.method} ${req.route.path}`);
 ```

File: src/controllers/update/check.your.answers.controller.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+    const appData: ApplicationData = getApplicationData(req.session);
+
+    return res.render(UPDATE_CHECK_YOUR_ANSWERS_PAGE, {
+      backLinkUrl: OVERSEAS_ENTITY_QUERY_URL, // TO DO: update whenever UAR-106 merged
+      templateName: UPDATE_CHECK_YOUR_ANSWERS_PAGE,
+      appData
```
Comment (dgolubevs) [inline src/controllers/update/check.your.answers.controller.ts:28]: why need pass `appData` 
i did not found any usage in template

File: src/controllers/update/check.your.answers.controller.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    const session = req.session as Session;
+
+    const transactionID = await postTransaction(req, session);
+    const overseasEntityID = await createOverseasEntity(req, session, transactionID);
+
+    const transactionClosedResponse = await closeTransaction(req, session, transactionID, overseasEntityID);
+    logger.infoRequest(req, `Transaction Closed, ID: ${transactionID}`);
+
+    const baseURL = `${CHS_URL}${UPDATE_AN_OVERSEAS_ENTITY_URL}`;
+    const redirectPath = await startPaymentsSession(req, session, transactionID, overseasEntityID, transactionClosedResponse, baseURL);
```
Comment (dgolubevs) [inline src/controllers/update/check.your.answers.controller.ts:49]: line too long, more readable
```
const redirectPath = await startPaymentsSession(
  req, 
  session, 
  transactionID,
  overseasEntityID,
  transactionClosedResponse
  baseURL,
);

File: src/service/payment.service.ts [hunk 11]
```diff
@@ -24,7 +24,7 @@ import { OverseasEntityKey, PaymentKey, Transactionkey } from "../model/data.typ
 // directing the application to the Payment Platform to begin a payment session, otherwise
 // will return the CONFIRMATION URL.
 export const startPaymentsSession = async (
-  req: Request, session: Session, transactionId: string, overseasEntityId: string, transactionRes
+  req: Request, session: Session, transactionId: string, overseasEntityId: string, transactionRes, baseURL?: string
```
Comment (dgolubevs) [inline src/service/payment.service.ts:27]: line too long
```
 = async (
  req: Request,
  session: Session,
  transactionId: string,
  overseasEntityId: string,
  transactionRes,
  baseURL?: string,
): Promise...
```

File: src/controllers/update/check.your.answers.controller.ts (hunk 8 above)
Comment (corybudd-ch) [reply src/controllers/update/check.your.answers.controller.ts]: thank you, changed now

File: src/controllers/update/check.your.answers.controller.ts (hunk 10 above)
Comment (corybudd-ch) [reply src/controllers/update/check.your.answers.controller.ts]: changed

File: src/service/payment.service.ts (hunk 11 above)
Comment (corybudd-ch) [reply src/service/payment.service.ts]: changed

File: src/controllers/update/check.your.answers.controller.ts (hunk 9 above)
Comment (dgolubevs) [reply src/controllers/update/check.your.answers.controller.ts:28]: On current moment in `views/update/update-check-your-answers.html` you do not use any `addData.<field>`
so there is no reason to pass do not used object to template.
better do not keep do not used source code

File: src/routes/index.ts [hunk 12]
```diff
@@ -191,7 +192,11 @@ router.post(config.OVERSEAS_ENTITY_QUERY_URL, authentication, ...validator.overs
 router.get(config.UPDATE_OVERSEAS_ENTITY_CONFIRM_URL, authentication, confirmOverseasEntityDetails.get);
 router.post(config.UPDATE_OVERSEAS_ENTITY_CONFIRM_URL, authentication, confirmOverseasEntityDetails.post);
 router.get(config.OVERSEAS_ENTITY_REVIEW_URL, authentication, overseasEntityReview.get);
+router.post(config.OVERSEAS_ENTITY_REVIEW_URL, authentication, overseasEntityReview.post);
```
Comment (dgolubevs) [inline src/routes/index.ts:195]: better less duplication
```
router.route(config.OVERSEAS_ENTITY_REVIEW_URL)
   .all( authentication)
   .get(overseasEntityReview.get)
   .post( overseasEntityReview.post);

File: src/routes/index.ts [hunk 13]
```diff
@@ -191,7 +192,11 @@ router.post(config.OVERSEAS_ENTITY_QUERY_URL, authentication, ...validator.overs
...
 router.get(config.OVERSEAS_ENTITY_UPDATE_DETAILS_URL, authentication, overseasEntityUpdateDetails.get);
 router.post(config.OVERSEAS_ENTITY_UPDATE_DETAILS_URL, authentication, ...validator.entity, checkValidations, overseasEntityUpdateDetails.post);
 
+router.get(config.UPDATE_CHECK_YOUR_ANSWERS_URL, authentication, updateCheckYourAnswers.get);
```
Comment (dgolubevs) [inline src/routes/index.ts:199]:
```
router.route(config. UPDATE_CHECK_YOUR_ANSWERS_URL)
   .all( authentication)
   .get(updateCheckYourAnswers.get)
   .post(updateCheckYourAnswers.post);
```

File: test/controllers/update/check.your.answers.controller.spec.ts [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+mockOverseasEntity.mockReturnValue( OVERSEAS_ENTITY_ID );
+
+const mockCloseTransaction = closeTransaction as jest.Mock;
+mockCloseTransaction.mockReturnValue( TRANSACTION_CLOSED_RESPONSE );
+
+const mockPaymentsSession = startPaymentsSession as jest.Mock;
+mockPaymentsSession.mockReturnValue( "CONFIRMATION_URL" );
+
+describe("CHECK YOUR ANSWERS controller", () => {
+
```
Comment (dgolubevs) [inline test/controllers/update/check.your.answers.controller.spec.ts:59]: for consistency better do not leave empty line after '{' or before `}` in one please and do not add in other

```
describe("GET tests", () => {
-> added new line
    test(`renders the ${UPDATE_CHECK_YOUR_ANSWERS_PAGE} page`, async () => {
    -> not added new line
```

**Review by dgolubevs — APPROVED** [review]

File: test/controllers/update/check.your.answers.controller.spec.ts (hunk 14 above)
Comment (corybudd-ch) [reply test/controllers/update/check.your.answers.controller.spec.ts]: fixed

File: src/routes/index.ts (hunk 13 above)
Comment (corybudd-ch) [reply src/routes/index.ts]: fixed

File: src/routes/index.ts (hunk 12 above)
Comment (corybudd-ch) [reply src/routes/index.ts]: fixed

File: src/controllers/update/check.your.answers.controller.ts (hunk 9 above)
Comment (corybudd-ch) [reply src/controllers/update/check.your.answers.controller.ts]: done
